data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaSinkProducer.java

```diff
@@ -0,0 +1,196 @@
+        this.dlqSink = dlqSink;
+        this.schemaRegistryClient = schemaRegistryClient;
+        bufferedEventHandles = new LinkedList<>();
+        this.expressionEvaluator = expressionEvaluator;
+        this.tagTargetKey = tagTargetKey;
+
+    }
+
+    public void produceRecords(final Record<Event> record) {
+        if (record.getData().getEventHandle() != null) {
+            bufferedEventHandles.add(record.getData().getEventHandle());
+        }
+        TopicConfig topic = kafkaSinkConfig.getTopic();
+        Event event = getEvent(record);
+        final String key = event.formatString(kafkaSinkConfig.getPartitionKey(), expressionEvaluator);
+        Object dataForDlq = event.toJsonString();
+        LOG.info("Producing record " + dataForDlq);
+        try {
+            final String serdeFormat = kafkaSinkConfig.getSerdeFormat();
+            if (MessageFormat.JSON.toString().equalsIgnoreCase(serdeFormat)) {
+                publishJsonMessage(record, topic, key, dataForDlq);
+            } else if (MessageFormat.AVRO.toString().equalsIgnoreCase(serdeFormat)) {
+                publishAvroMessage(record, topic, key, dataForDlq);
+            } else {
+                publishPlaintextMessage(record, topic, key, dataForDlq);
+            }
+        } catch (Exception e) {
+            releaseEventHandles(false);
+        }
+
+    }
+
+    private Event getEvent(Record<Event> record) {
+        Event event = record.getData();
+        try {
+            event = addTagsToEvent(event, tagTargetKey);
+        } catch (JsonProcessingException e) {
+            LOG.error("error occured while processing tag target key");
+        }
+        return event;
+    }
+
+    private void publishPlaintextMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) {
+        producer.send(new ProducerRecord(topic.getName(), key, record.getData().toJsonString()), callBack(dataForDlq));
+    }
+
+    private void publishAvroMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) throws RestClientException, IOException {
+        final String valueToParse = schemaRegistryClient.
+                getLatestSchemaMetadata(topic.getName() + "-value").getSchema();
+        final Schema schema = new Schema.Parser().parse(valueToParse);
+        final GenericRecord genericRecord = getGenericRecord(record.getData(), schema);
+        producer.send(new ProducerRecord(topic.getName(), key, genericRecord), callBack(dataForDlq));
+    }
+
+    private void publishJsonMessage(Record<Event> record, TopicConfig topic, String key, Object dataForDlq) throws IOException, RestClientException, ProcessingException {
+        final JsonNode dataNode = new ObjectMapper().convertValue(record.getData().toJsonString(), JsonNode.class);
+        if (validateJson(topic.getName(), dataForDlq)) {
+            producer.send(new ProducerRecord(topic.getName(), key, dataNode), callBack(dataForDlq));
```

Comment:
producer.send is repeated in all functions. functions can be refarctoed to getJsonMessage etc and producer.send at one place

Reply:
we will incorporate this change in next PR.